### ISSUE TITLE
Compilation issue with Visual Studio v16.8.0 and higher

### ISSUE THREAD

**Issue by spindensity** [pytorch/pytorch#48517]
# Summary

Moving from Visual Studio v16.7.7 to v16.8.2, we now see the following error when compiling pytorch 1.7.0 or the master branch:

```log
caffe2\CMakeFiles\torch_cpu.dir\__\aten\src\ATen\native\cpu\SortingKernel.cpp.DEFAULT.cpp.obj
FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/SortingKernel.cpp.DEFAULT.cpp.obj
C:\PROGRA~2\MICROS~4\2019\ENTERP~1\VC\Tools\MSVC\1428~1.293\bin\Hostx64\x64\cl.exe  /nologo /TP -DCPUINFO_SUPPORTED_PLATFORM=1 -DFMT_HEADER_ONLY=1 -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DUSE_DISTRIBUTED -DUSE_EXTERNAL_MZCRC -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -Dtorch_cpu_EXPORTS -Iaten\src -I..\aten\src -I. -I..\ -I..\cmake\..\third_party\benchmark\include -Icaffe2\contrib\aten -I..\third_party\onnx -Ithird_party\onnx -I..\third_party\foxi -Ithird_party\foxi -I..\torch\csrc\api -I..\torch\csrc\api\include -I..\caffe2\aten\src\TH -Icaffe2\aten\src\TH -Icaffe2\aten\src -Icaffe2\..\aten\src -Icaffe2\..\aten\src\ATen -I..\torch\csrc -I..\third_party\miniz-2.0.8 -I..\aten\src\TH -I..\aten\..\third_party\catch\single_include -I..\aten\src\ATen\.. -Icaffe2\aten\src\ATen -I..\caffe2\core\nomnigraph\include -I..\c10\.. -Ithird_party\ideep\mkl-dnn\include -I..\third_party\ideep\mkl-dnn\src\..\include -I..\third_party\pthreadpool\include -I..\third_party\cpuinfo\include -I..\third_party\fbgemm\include -I..\third_party\fbgemm -I..\third_party\fbgemm\third_party\asmjit\src -I..\third_party\FP16\include -I..\third_party\fmt\include -Ithird_party\gloo -I..\cmake\..\third_party\gloo -I..\cmake\..\third_party\googletest\googlemock\include -I..\cmake\..\third_party\googletest\googletest\include -I..\third_party\protobuf\src -I..\third_party\XNNPACK\include -I..\third_party -I..\cmake\..\third_party\eigen -IC:\tools\Anaconda3\envs\pytorch-build-py36\include -IC:\tools\Anaconda3\envs\pytorch-build-py36\lib\site-packages\numpy\core\include -I..\cmake\..\third_party\pybind11\include -I..\cmake\..\third_party\cub -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include" -I..\third_party\ideep\mkl-dnn\include -I..\third_party\ideep\include -I..\caffe2 /DWIN32 /D_WINDOWS /GR /EHsc /w /bigobj -DUSE_PTHREADPOOL -openmp:experimental -DNDEBUG -DUSE_FBGEMM -DUSE_XNNPACK -DUSE_VULKAN_WRAPPER -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2 /DNDEBUG /w /bigobj -DNDEBUG -DCAFFE2_USE_GLOO -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /EHsc /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -openmp:experimental -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /fp:strict  /DCPU_CAPABILITY=DEFAULT /DCPU_CAPABILITY_DEFAULT /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\__\aten\src\ATen\native\cpu\SortingKernel.cpp.DEFAULT.cpp.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c aten\src\ATen\native\cpu\SortingKernel.cpp.DEFAULT.cpp
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7417): error C2678: binary '*': no operator found which takes a left-hand operand of type 'const _BidIt' (or there is no acceptable conversion)
        with
        [
            _BidIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<scalar_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>
        ]
..\aten\src\ATen/native/CompositeRandomAccessorCommon.h(132): note: could be 'at::native::references_holder<std::tuple<unsigned char,__int64>,std::tuple<unsigned char &,__int64 &>> at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<scalar_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>::operator *(void)'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7417): note: while trying to match the argument list '(const _BidIt)'
        with
        [
            _BidIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<scalar_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7541): note: see reference to function template instantiation '_BidIt std::_Insertion_sort_unchecked<_RanIt,_Pr>(const _BidIt,const _BidIt,_Pr)' being compiled
        with
        [
            _BidIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<scalar_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>,
            _RanIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<scalar_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>,
            _Pr=at::native::`anonymous-namespace'::KeyValueCompDesc<scalar_t>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7571): note: see reference to function template instantiation 'void std::_Sort_unchecked<_RanIt,_Fn>(_RanIt,_RanIt,__int64,_Pr)' being compiled
        with
        [
            _RanIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<scalar_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>,
            _Fn=at::native::`anonymous-namespace'::KeyValueCompDesc<scalar_t>,
            _Pr=at::native::`anonymous-namespace'::KeyValueCompDesc<scalar_t>
        ]
aten\src\ATen\native\cpu\SortingKernel.cpp.DEFAULT.cpp(120): note: see reference to function template instantiation 'void std::sort<at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<scalar_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>,at::native::`anonymous-namespace'::KeyValueCompDesc<scalar_t>>(const _RanIt,const _RanIt,_Pr)' being compiled
        with
        [
            _RanIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<scalar_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>,
            _Pr=at::native::`anonymous-namespace'::KeyValueCompDesc<scalar_t>
        ]
aten\src\ATen\native\cpu\SortingKernel.cpp.DEFAULT.cpp(51): note: see reference to function template instantiation 'void at::native::`anonymous-namespace'::sort_kernel::<lambda_034da3ae7e8f2581ff071e14a0d250e8>::operator ()<scalar_t,int64_t>(scalar_t *,int64_t,int64_t *,int64_t,int64_t) const' being compiled
aten\src\ATen\native\cpu\SortingKernel.cpp.DEFAULT.cpp(127): note: see reference to function template instantiation 'void at::native::`anonymous-namespace'::_dim_apply<at::native::`anonymous-namespace'::sort_kernel::<lambda_034da3ae7e8f2581ff071e14a0d250e8>>(at::Tensor &,at::Tensor &,int64_t,const std::string &,const func_t &)' being compiled
        with
        [
            func_t=at::native::`anonymous-namespace'::sort_kernel::<lambda_034da3ae7e8f2581ff071e14a0d250e8>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7417): error C2100: illegal indirection
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7419): error C2678: binary '*': no operator found which takes a left-hand operand of type 'const _BidIt' (or there is no acceptable conversion)
        with
        [
            _BidIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<scalar_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>
        ]
..\aten\src\ATen/native/CompositeRandomAccessorCommon.h(132): note: could be 'at::native::references_holder<std::tuple<unsigned char,__int64>,std::tuple<unsigned char &,__int64 &>> at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<scalar_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>::operator *(void)'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7419): note: while trying to match the argument list '(const _BidIt)'
        with
        [
            _BidIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<scalar_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7419): error C2100: illegal indirection
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7417): error C2678: binary '*': no operator found which takes a left-hand operand of type 'const _BidIt' (or there is no acceptable conversion)
        with
        [
            _BidIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>
        ]
..\aten\src\ATen/native/CompositeRandomAccessorCommon.h(132): note: could be 'at::native::references_holder<std::tuple<__int64,__int64>,std::tuple<__int64 &,__int64 &>> at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>::operator *(void)'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7417): note: while trying to match the argument list '(const _BidIt)'
        with
        [
            _BidIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7541): note: see reference to function template instantiation '_BidIt std::_Insertion_sort_unchecked<_RanIt,_Pr>(const _BidIt,const _BidIt,_Pr)' being compiled
        with
        [
            _BidIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>,
            _RanIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>,
            _Pr=at::native::`anonymous-namespace'::KeyValueCompDesc<scalar_t>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7571): note: see reference to function template instantiation 'void std::_Sort_unchecked<_RanIt,_Fn>(_RanIt,_RanIt,__int64,_Pr)' being compiled
        with
        [
            _RanIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>,
            _Fn=at::native::`anonymous-namespace'::KeyValueCompDesc<scalar_t>,
            _Pr=at::native::`anonymous-namespace'::KeyValueCompDesc<scalar_t>
        ]
aten\src\ATen\native\cpu\SortingKernel.cpp.DEFAULT.cpp(120): note: see reference to function template instantiation 'void std::sort<at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>,at::native::`anonymous-namespace'::KeyValueCompDesc<scalar_t>>(const _RanIt,const _RanIt,_Pr)' being compiled
        with
        [
            _RanIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>,
            _Pr=at::native::`anonymous-namespace'::KeyValueCompDesc<scalar_t>
        ]
aten\src\ATen\native\cpu\SortingKernel.cpp.DEFAULT.cpp(51): note: see reference to function template instantiation 'void at::native::`anonymous-namespace'::sort_kernel::<lambda_034da3ae7e8f2581ff071e14a0d250e8>::operator ()<scalar_t,int64_t>(scalar_t *,int64_t,int64_t *,int64_t,int64_t) const' being compiled
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7419): error C2678: binary '*': no operator found which takes a left-hand operand of type 'const _BidIt' (or there is no acceptable conversion)
        with
        [
            _BidIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>
        ]
..\aten\src\ATen/native/CompositeRandomAccessorCommon.h(132): note: could be 'at::native::references_holder<std::tuple<__int64,__int64>,std::tuple<__int64 &,__int64 &>> at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>::operator *(void)'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\include\algorithm(7419): note: while trying to match the argument list '(const _BidIt)'
        with
        [
            _BidIt=at::native::CompositeRandomAccessor<at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::StridedRandomAccessor<int64_t,int64_t,at::native::DefaultPtrTraits>,at::native::TupleInfoCPU>
        ]
```

# Environment

* CUDA: 11.0;
* cuDNN: 8.0.5;
* OS version: Windows 10 20H2 64bit;
* Compiler: MSVC 16.8.0 or higher;
* CMake: 3.18.4;

# Additional information

Someone has reported it to Microsoft, the discussion is [here](https://developercommunity.visualstudio.com/content/problem/1255147/vs-1680-clexe-regression-causing-pytorch-build-fai.html).

cc @malfet @seemethere @walterddr @peterjc123 @maxluk @nbcsm @guyang3532 @gunandrose4u @mszhanyi @skyline75489

**Comment by cheerser** [comment]
same problem

**Comment by peterjc123** [comment]
@ezyang Any ideas?

**Comment by malfet** [comment]
This sounds interesting, tentatively grabbing for myself...

**Comment by skyline75489** [comment]
I’m the one that reported the issue in VS community. It works with VS 16.7 so I think it’s a regression in the compiler itself.

**Comment by malfet** [comment]
This compilation failure can be fixed by adding a const qualifier to `CompositeRandomAccessor::operator*()`

**Comment by malfet** [comment]
Closing, as it is already fixed by #47853
